### PR TITLE
DON-723: Create update payment method modal

### PR DESCRIPTION
**Pull request by bdsl**
This creates a new modal component for updating a card payment method.

There's more work to do, but the button for launching this modal is set to only show in the dev environment, so this should be safe to merge as-is, subject to review.

image

image

### CODE REVIEW

File: src/app/update-card-modal/update-card-modal.component.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    MatButtonModule,
+    MatDialogModule,
+    MatInputModule,
+    MatProgressSpinnerModule,
+    ReactiveFormsModule,
+    RecaptchaModule,
+  ],
+})
+export class UpdateCardModalComponent implements OnInit {
+  @ViewChild('captcha') captcha: RecaptchaComponent;
```
Comment (bdsl) [inline src/app/update-card-modal/update-card-modal.component.ts:30]: I think we can get rid of the captcha - we're already logged in and have a payment card saved.

File: src/app/update-card-modal/update-card-modal.html [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+<!--
+  *ngIf="card" above shouldn't be needed - we don't load the modal at all without a card, but it makes the `it('should create'`
+  test pass. Would be good to remove.
+ -->
+    <h2 mat-dialog-title>Update Payment Card </h2>
+
+  <mat-dialog-content class="mat-typography">
+      <form [formGroup]="updateCardForm">
+        <mat-form-field color="primary">
+          <mat-label for="expiryDate">Brand</mat-label>
```
Comment (bdsl) [inline src/app/update-card-modal/update-card-modal.html:11]: label above is wrong

File: src/app/update-card-modal/update-card-modal.html [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+              id="expiryDate"
+              formControlName="expiryDate"
+            >
+          </mat-form-field>
+
+        <mat-form-field color="primary">
+          <mat-label for="billingAddress">Billing Address</mat-label>
+
+          <!--
+          @todo Prefil with old billing address if possible, work out what format we should present the address in.
```
Comment (NoelLH) [inline src/app/update-card-modal/update-card-modal.html:56]: I think that as we discussed last week, postcode and country is probably sufficient for how we currently send + use this? Sounds like that's what we're going to split up from discussion today.

File: src/app/update-card-modal/update-card-modal.html (hunk 2 above)
Comment (bdsl) [reply src/app/update-card-modal/update-card-modal.html:11]: (actually the label was ok but the title and for attribute were wrong)

File: src/app/update-card-modal/update-card-modal.html [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          <mat-label for="expiryDate">Brand</mat-label>
+          <input
+            [disabled]=true
+            width="5"
+            matInput
+            type="text"
+            title="Brand"
+            [value]="card.brand.toUpperCase()"
+            id="brand"
+            formControlName="last4"
```
Comment (bdsl) [inline src/app/update-card-modal/update-card-modal.html:20]:
```suggestion
  <mat-dialog-content class="mat-typography">
      <form [formGroup]="updateCardForm">
        <mat-form-field color="primary">
          <mat-label for="brand">Brand</mat-label>
          <input
            [disabled]=true
            width="5"
            matInput
            type="text"
            title="Brand"
            [value]="card.brand.toUpperCase()"
            id="brand"
            formControlName="brand"
```

File: src/app/update-card-modal/update-card-modal.html [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+          >
+        </mat-form-field>
+
+        <mat-form-field color="primary">
+          <mat-label for="expiryDate">Last Four Digits</mat-label>
+          <input
```
Comment (bdsl) [inline src/app/update-card-modal/update-card-modal.html:26]:
```suggestion
        <mat-form-field color="primary">
          <mat-label for="last4">Last Four Digits</mat-label>
          <input
```

File: src/app/update-card-modal/update-card-modal.html [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+          <!--
+          @todo Prefil with old billing address if possible, work out what format we should present the address in.
+                Probably replace with multiple form fields for the individual properties of the address.
+          -->
+          <textarea
+            matInput
+            title="Billing Address"
+            [value]=""
+            id="Billing Address"
+            formControlName="expiryDate"
```
Comment (bdsl) [inline src/app/update-card-modal/update-card-modal.html:64]:
```suggestion
            formControlName="billingAddress"
```

File: src/app/update-card-modal/update-card-modal.html [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+          <!--
+          @todo Prefil with old billing address if possible, work out what format we should present the address in.
+                Probably replace with multiple form fields for the individual properties of the address.
+          -->
+          <textarea
+            matInput
+            title="Billing Address"
+            [value]=""
+            id="Billing Address"
```
Comment (NoelLH) [inline src/app/update-card-modal/update-card-modal.html:63]:
```suggestion
            id="billingAddress"
```

**Review by NoelLH — APPROVED** [review]

File: src/app/update-card-modal/update-card-modal.html (hunk 3 above)
Comment (bdsl) [reply src/app/update-card-modal/update-card-modal.html:56]: Yes that's the plan.

**Pull request merged**